vowpalwabbit/automl.h

```diff
@@ -0,0 +1,173 @@
+  size_t config_index = 0;
+  bool eligible_to_inactivate = false;
+  interaction_vec_t live_interactions;  // Live pre-allocated vectors in use
+
+  scored_config() : chisq(0.05, 0.999, 0, std::numeric_limits<double>::infinity()) {}
+
+  void update(float w, float r);
+  void save_load(io_buf&, bool, bool);
+  void persist(metric_sink&, const std::string&);
+  float current_ips() const;
+  void reset_stats();
+};
+
+// all possible states of exclusion config
+enum class config_state
+{
+  New,
+  Live,
+  Inactive,
+  Removed
+};
+
+struct exclusion_config
+{
+  std::map<namespace_index, std::set<namespace_index>> exclusions;
+  size_t lease;
+  float ips = 0.f;
+  float lower_bound = std::numeric_limits<float>::infinity();
+  config_state state = VW::automl::config_state::New;
+
+  exclusion_config(size_t lease = 10) : lease(lease) {}
+
+  void save_load(io_buf&, bool, bool);
+};
+
+// all possible states of automl
+enum class automl_state
+{
+  Collecting,
+  Experimenting
+};
+
+struct config_manager
+{
+  // This fn is responsible for applying a config
+  // tracked by 'live_slot' into the example.
+  // the impl is responsible of tracking this config-live_slot mapping
+  void apply_config(example*, size_t);
+  // This fn is the 'undo' of configure_interactions
+  void revert_config(example*);
+  void save_load(io_buf&, bool, bool);
+  void persist(metric_sink&);
+
+  // Public Chacha functions
+  void config_oracle();
+  void pre_process(const multi_ex&);
+  void schedule();
+  void update_champ();
+};
+
+using priority_func = float(const exclusion_config&, const std::map<namespace_index, size_t>&);
+
+struct interaction_config_manager : config_manager
+{
+  size_t total_learn_count = 0;
+  size_t current_champ = 0;
+  size_t global_lease;
+  size_t max_live_configs;
+  uint64_t seed;
+  rand_state random_state;
+  size_t priority_challengers;
+
+  // Stores all namespaces currently seen -- Namespace switch could we use array, ask Jack
+  std::map<namespace_index, size_t> ns_counter;
+
+  // Stores all configs in consideration (Map allows easy deletion unlike vector)
+  std::map<size_t, exclusion_config> configs;
+
+  // Stores scores of live configs, size will never exceed max_live_configs
+  std::vector<scored_config> scores;
+
+  // Maybe not needed with oracle, maps priority to config index, unused configs
+  std::priority_queue<std::pair<float, size_t>> index_queue;
+
+  interaction_config_manager(
+      size_t, size_t, uint64_t, size_t, float (*)(const exclusion_config&, const std::map<namespace_index, size_t>&));
+
+  void apply_config(example*, size_t);
+  void revert_config(example*);
+  void save_load(io_buf&, bool, bool);
+  void persist(metric_sink&);
+
+  // Public Chacha functions
+  void config_oracle();
+  void pre_process(const multi_ex&);
+  void schedule();
+  void update_champ();
+
+private:
+  void gen_quadratic_interactions(size_t);
+  bool better(const exclusion_config&, const exclusion_config&) const;
```

Comment:
Toying with the idea of making better, worse independent of config_manager implentation.... 

If better & worse accepted a class representing model performance "model_stats", then better or worse can be independent of config_manager.
If that were the case, current better() fn could be implmented outside the config_manager and be called better_lower_bound(stat1, stat2) for example.

Reply:
I think that makes sense but I'm concerned its preemptive. This would require us to template the automl, config_manager, and exclusion_config structs, as well as many functions within them, and up to 3 new flags for stats_type, better_func, worse_func (which may or may not be relevant in the future). This may overcomplicate things depending on how automl gets configured in the future